File: 백준/Gold/2294. 동전 2/동전 2.cc
#include <iostream>

using namespace std;

const int MX = 100002;
int coins[102];
int dp[MX];

int main(){
    int n, k;
    cin >> n >> k;
    
    for(int i = 0; i < n; i++){
        cin >> coins[i];
    }
    fill(dp, dp+MX, 1e9);
    dp[0] = 0;
    for(int i = 0; i < n; i++){
        int coin = coins[i];
        for(int j = coin; j <= 10000; j++){
            dp[j] = min(dp[j-coin] + 1, dp[j]);
        }
    }
    if(dp[k] == 1e9) cout << -1;
    else cout << dp[k];

    return 0;
}

File: 백준/Gold/2294. 동전 2/동전 2.java
import java.io.*;
import java.util.*;

public class Main {

	public static void main(String[] args) throws Exception {
		// System.setIn(new FileInputStream("src/input.txt"));
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		int k = sc.nextInt();

		int coins[] = new int[n];

		for (int i = 0; i < n; i++) {
			coins[i] = sc.nextInt();
		}

		int dp[] = new int[k + 1];

		Arrays.fill(dp, (int) 1e9);
		dp[0] = 0;
		for (int coin : coins) {
			for (int i = coin; i <= k; i++) {
				dp[i] = Math.min(dp[i], dp[i - coin] + 1);
			}
//			System.out.println(Arrays.toString(dp));
		}

		if (dp[k] == (int) 1e9) {
			System.out.println(-1);
			return;
		}
		System.out.println(dp[k]);
	}

}